### PR TITLE
[prometheus-pushgateway] Enhancement for multiple containers, support of secrets and authenticated ServiceMonitor

### PR DESCRIPTION
**Pull request by pburgisser**
Signed-off-by: Philippe Bürgisser <philippe.burgisser@camptocamp.com>

#### What this PR does / why we need it:
- Adding support of multiple container in the same pod (side car, like oAuth-proxy)
- Adding support of multiple secret (necessary for oauth)
- Adding support of annotation on service, service account
- Adding support of authenticated Service Monitor

- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

File: charts/prometheus-pushgateway/Chart.yaml [hunk 1]
```diff
@@ -2,7 +2,7 @@ apiVersion: v1
 appVersion: "1.2.0"
 description: A Helm chart for prometheus pushgateway
 name: prometheus-pushgateway
-version: 1.4.2
+version: 1.4.3
```
Comment (monotek) [inline charts/prometheus-pushgateway/Chart.yaml:5]: New features should raise minorv version, not bugfix -> 1.5.0

File: charts/prometheus-pushgateway/templates/deployment.yaml [hunk 2]
```diff
@@ -24,6 +24,7 @@ spec:
     spec:
       serviceAccountName: {{ template "prometheus-pushgateway.serviceAccountName" . }}
       containers:
+{{ toYaml .Values.containers | indent 8}}
```
Comment (monotek) [inline charts/prometheus-pushgateway/templates/deployment.yaml:27]: imho should be name "extraContainers".
also schould have an example in values.yaml.

File: charts/prometheus-pushgateway/templates/deployment.yaml [hunk 3]
```diff
@@ -71,13 +73,21 @@ spec:
...
 {{ toYaml .Values.affinity | indent 8 }}
     {{- end }}
-      {{- if .Values.persistentVolume.enabled }}
+      {{- if or .Values.persistentVolume.enabled .Values.secrets }}
     {{- if .Values.securityContext }}
       securityContext:
 {{ toYaml .Values.securityContext | indent 8 }}
     {{- end }}
       volumes:
+        {{- range .Values.secrets }}
```
Comment (monotek) [inline charts/prometheus-pushgateway/templates/deployment.yaml:82]: Should be named "extraVolumes" and have an example in the values file too.

**Review by monotek — CHANGES_REQUESTED** [review]

File: charts/prometheus-pushgateway/Chart.yaml (hunk 1 above)
Comment (pburgisser-c2c) [reply charts/prometheus-pushgateway/Chart.yaml:5]: Ok thanks for this!

File: charts/prometheus-pushgateway/templates/deployment.yaml (hunk 2 above)
Comment (pburgisser-c2c) [reply charts/prometheus-pushgateway/templates/deployment.yaml:27]: No problem. I tried to be as close a possible from the naming used in the prometheus-operator. But I'll do as you said

File: charts/prometheus-pushgateway/templates/deployment.yaml (hunk 3 above)
Comment (pburgisser-c2c) [reply charts/prometheus-pushgateway/templates/deployment.yaml:82]: Hmmmm, imho, I would like to keep it this way as  mounting a secret is different from mounting other kind of storage. Unless we drop the idea of letting the user to add secret and let him add whatever needed as a yaml value. We could even add the two options actually (secret or a raw yaml value) What do you think @monotek ?

**Comment by pburgisser-c2c** [comment]
Hmmmm, I was not very familiar with that signoff and when running the command 
`git rebase HEAD~15 --signoff && git push --force-with-lease origin pushgateway_enhancement
` as indicated in the github action, it suddenly added all the people to review. Sorry for the noise guys.

File: charts/prometheus-pushgateway/templates/deployment.yaml (hunk 3 above)
Comment (monotek) [reply charts/prometheus-pushgateway/templates/deployment.yaml:84]: But with this function you could add genric volumes and therefore i'd like to see a generic name there.

Would be also mor consistent with prometheus chart. See: https://github.com/prometheus-community/helm-charts/blob/main/charts/prometheus/values.yaml#L650-L687

Comment (pburgisser-c2c) [reply charts/prometheus-pushgateway/templates/deployment.yaml]: There you go :-)

**Review by monotek — CHANGES_REQUESTED** [review]
please remove the alertmanager changes.

**Comment by pburgisser-c2c** [comment]
But the lint check fails due to the version :-/

**Comment by monotek** [comment]
Yes, i guess because you have changes for 2 charts here.
So please remove the alertmanager changes:
https://github.com/prometheus-community/helm-charts/pull/178/files#diff-03062d9d03647a312a91676379453d7c
https://github.com/prometheus-community/helm-charts/pull/178/files#diff-e2a9a571c3be079f7d9fae2912da24bc

**Comment by pburgisser-c2c** [comment]
Thanks @monotek ! I reverted it.
The code for Pushgateway which the Lint Code Base used to succeed now fails on kubeval. Is that kubeval something new? Thanks

**Review by monotek — APPROVED** [review]

**Comment by monotek** [comment]
Kubeval should only test kubernetes manifests. 
Not sure why it tries to lint the Chart.yaml.

@torstenwalter @scottrigby 
I see no recent change to the github actions workflow and the superlinter is fixed to v3 so why is this popping up now?
I also can't find anything in the superlinter repo or the used garethr/kubeval docker container which could do a "helm template" to create the manifests.
Sure the kubeval tests work for helm charts?

**Comment by torstenwalter** [comment]
@monotek Create a PR to fix it: https://github.com/prometheus-community/helm-charts/pull/191

**Comment by monotek** [comment]
@torstenwalter 
Thanks! :)
See my comment in https://github.com/prometheus-community/helm-charts/pull/191#issuecomment-705594777.

@pburgisser-c2c 

Could you merge the changes from master to your branch introduced by #191 please.
After that linting should succeed :)

**Comment by pburgisser-c2c** [comment]
@monotek Ok, it worked ! Thanks!

**Review by monotek — APPROVED** [review]

**Comment by pburgisser-c2c** [comment]
cc @gianrubio For review :-)
Thanks for your help

**Comment by monotek** [comment]
@gianrubio please review :)

**Comment by pburgisser** [comment]
Hello @gianrubio,

May I ask you to have a look please :-)

Thanks and best regards
Philippe

File: charts/prometheus-pushgateway/Chart.yaml [hunk 4]
```diff
@@ -2,7 +2,7 @@ apiVersion: v1
 appVersion: "1.2.0"
 description: A Helm chart for prometheus pushgateway
 name: prometheus-pushgateway
-version: 1.4.2
+version: 1.5.0
```
Comment (scottrigby) [inline charts/prometheus-pushgateway/Chart.yaml:5]: Needs a MINOR bump again. See https://github.com/prometheus-community/helm-charts/blob/main/charts/prometheus-pushgateway/Chart.yaml

Related: we should probably implement https://github.com/prometheus-community/helm-charts/issues/249 sooner than later. Anyone else want to weigh in on that?

**Review by scottrigby — CHANGES_REQUESTED** [review]
Otherwise lgtm

**Comment by pburgisser** [comment]
@scottrigby The source branch has been deleted (I made the mistake).
I could recover from this PR and made the change you requested, unfortunately and according to GH docs, I had to do a new PR [1]
[1] https://github.com/prometheus-community/helm-charts/pull/493/

**Comment by pburgisser** [comment]
Moved to https://github.com/prometheus-community/helm-charts/pull/493

**Pull request closed without merging**